File: cmd/remove/repo.go
package remove

import (
	"fmt"

	"github.com/fioncat/roxide/cmd"
	"github.com/fioncat/roxide/pkg/choice"
	"github.com/fioncat/roxide/pkg/context"
	"github.com/fioncat/roxide/pkg/db"
	"github.com/fioncat/roxide/pkg/repoutils"
	"github.com/fioncat/roxide/pkg/term"
	"github.com/spf13/cobra"
)

func newRepo() *cobra.Command {
	var opts repoOptions

	c := &cobra.Command{
		Use:   "repo [HEAD] [QUERY]",
		Short: "Remove one or more repositories",

		Args: cobra.MaximumNArgs(2),

		ValidArgsFunction: cmd.BuildCompletion(cmd.RepoCompletion),
	}

	c.Flags().BoolVarP(&opts.recursive, "recursive", "r", false, "remove multiple repositories")
	c.Flags().BoolVarP(&opts.force, "force", "f", false, "force remove, ignore pin flag")

	return cmd.Build(c, &opts)
}

type repoOptions struct {
	args []string

	recursive bool

	force bool
}

func (o *repoOptions) Complete(c *cobra.Command, args []string) error {
	o.args = args
	return nil
}

func (o *repoOptions) Run(ctx *context.Context) error {
	if o.recursive {
		return o.runMany(ctx)
	} else {
		return o.runOne(ctx)
	}
}

func (o *repoOptions) runMany(ctx *context.Context) error {
	ch := choice.New(ctx, o.args)

	var opts choice.ManyOptions
	if !o.force {
		opts.Sync = db.BoolPtr(false)
	}
	list, err := ch.ManyLocal(opts)
	if err != nil {
		return err
	}
	if len(list.Items) == 0 {
		term.PrintInfo("No repo to remove")
		return nil
	}

	names := make([]string, 0, len(list.Items))
	for _, repo := range list.Items {
		names = append(names, repo.String())
	}

	err = term.ConfirmItems(names, "remove", "removal", "Repo", "Repos")
	if err != nil {
		return err
	}

	for _, repo := range list.Items {
		err = repoutils.Remove(ctx, repo)
		if err != nil {
			return fmt.Errorf("failed to remove repo %q: %w", repo.String(), err)
		}
	}

	return nil
}

func (o *repoOptions) runOne(ctx *context.Context) error {
	ch := choice.New(ctx, o.args)

	repo, err := ch.One(choice.OneOptions{
		Mode:       choice.ModeSelect,
		ForceLocal: true,
	})
	if err != nil {
		return err
	}

	return repoutils.Remove(ctx, repo)
}

File: src/cmd/remove/repo.rs
use anyhow::Result;
use async_trait::async_trait;
use clap::Args;

use crate::cmd::Command;
use crate::cmd::complete;
use crate::config::context::ConfigContext;
use crate::db::repo::Repository;
use crate::repo::ops::RepoOperator;
use crate::repo::select::SelectRepoArgs;
use crate::repo::select::{RepoSelector, SelectManyReposOptions};
use crate::term::confirm::confirm_items;
use crate::{confirm, debug, info};

#[derive(Debug, Args)]
pub struct RemoveRepoCommand {
    #[clap(flatten)]
    pub select_repo: SelectRepoArgs,

    #[arg(long, short)]
    pub recursive: bool,

    #[arg(long, short)]
    pub force: bool,
}
#[async_trait]
impl Command for RemoveRepoCommand {
    async fn run(self, ctx: ConfigContext) -> Result<()> {
        debug!("[cmd] Run remove repo command: {:?}", self);
        ctx.lock()?;

        let selector = RepoSelector::new(&ctx, &self.select_repo);
        if !self.recursive {
            let repo = selector.select_one(false, true).await?;
            confirm!("Are you sure to remove repository {}", repo.full_name());
            return self.remove(&ctx, repo);
        }

        let mut opts = SelectManyReposOptions::default();
        if !self.force {
            opts.pin = Some(false);
        }
        let list = selector.select_many(opts)?;
        if list.items.is_empty() {
            info!("No repo to remove");
            return Ok(());
        }

        let names = list.display_names();
        confirm_items(&names, "remove", "removal", "Repo", "Repos")?;

        for repo in list.items {
            self.remove(&ctx, repo)?;
        }

        Ok(())
    }

    fn complete_command() -> clap::Command {
        clap::Command::new("repo").args(complete::repo_args())
    }
}

impl RemoveRepoCommand {
    fn remove(&self, ctx: &ConfigContext, repo: Repository) -> Result<()> {
        let db = ctx.get_db()?;
        let op = RepoOperator::load(ctx, &repo)?;
        op.remove()?;
        db.with_transaction(|tx| tx.repo().delete(&repo))?;
        info!(
            "Repository {} was removed from disk and database",
            repo.full_name()
        );
        Ok(())
    }
}
